File: cpp/src/builder.cpp
// builder.cpp
#include <iostream>
#include <string>

// 계산기 클래스 (메소드 체이닝)
class Calculator {
private:
    double result;

public:
    Calculator() : result(0) {}

    Calculator& add(double value) {
        result += value;
        return *this;
    }

    Calculator& subtract(double value) {
        result -= value;
        return *this;
    }

    Calculator& multiply(double value) {
        result *= value;
        return *this;
    }

    Calculator& divide(double value) {
        if (value != 0) {
            result /= value;
        }
        return *this;
    }

    double getResult() const {
        return result;
    }
};

// Person 클래스 (빌더 패턴 + 메소드 체이닝)
class Person {
public:
    std::string name;
    int age;
    std::string address;
    
    void display() const {
        std::cout << "Name: " << name << ", Age: " << age << ", Address: " << address << std::endl;
    }
};

class PersonBuilder {
private:
    Person person;

public:
    PersonBuilder& withName(const std::string& name) {
        person.name = name;
        return *this;
    }

    PersonBuilder& withAge(int age) {
        person.age = age;
        return *this;
    }

    PersonBuilder& withAddress(const std::string& address) {
        person.address = address;
        return *this;
    }

    Person build() const {
        return person;
    }
};

// 테스트 코드
int main() {
    // 계산기 예제 - 메소드 체이닝
    Calculator calc;
    double result = calc.add(5).multiply(2).subtract(3).divide(2).getResult();
    std::cout << "Calculator result: " << result << std::endl;

    // Person 빌더 예제
    Person person = PersonBuilder()
        .withName("홍길동")
        .withAge(20)
        .withAddress("양산시 부산대학로")
        .build();
    
    person.display();
    
    return 0;
}

File: kotlin/src/main/kotlin/Builder.kt
// Builder.kt
package designpatterns

// 계산기 클래스 (메소드 체이닝)
class Calculator {
    private var result = 0.0

    fun add(value: Double): Calculator {
        result += value
        return this
    }

    fun subtract(value: Double): Calculator {
        result -= value
        return this
    }

    fun multiply(value: Double): Calculator {
        result *= value
        return this
    }

    fun divide(value: Double): Calculator {
        if (value != 0.0) {
            result /= value
        }
        return this
    }

    fun getResult(): Double {
        return result
    }
}

// Person 클래스 (빌더 패턴 + 메소드 체이닝)
class Person {
    var name: String = ""
    var age: Int = 0
    var address: String = ""
    
    fun display() {
        println("Name: $name, Age: $age, Address: $address")
    }

    // Kotlin에서는 companion object를 통해 정적 빌더 패턴을 구현할 수 있음
    class Builder {
        private val person = Person()

        fun withName(name: String): Builder {
            person.name = name
            return this
        }

        fun withAge(age: Int): Builder {
            person.age = age
            return this
        }

        fun withAddress(address: String): Builder {
            person.address = address
            return this
        }

        fun build(): Person {
            return person
        }
    }
}

// 테스트 함수
fun testBuilder() {
    // 계산기 예제 - 메소드 체이닝
    val calc = Calculator()
    val result = calc.add(5.0).multiply(2.0).subtract(3.0).divide(2.0).getResult()
    println("Calculator result: $result")

    // Person 빌더 예제
    val person = Person.Builder()
        .withName("홍길동")
        .withAge(20)
        .withAddress("양산시 부산대학로")
        .build()
    
    person.display()
}